### PR TITLE
Refactor plugin sync

### PR DESCRIPTION
**Pull request by iskhakov**
# What this PR does

This PR adds a shortcut in the plugin synchronisation process, so the existing users will be able login without waiting for the sync task. Every request still starts the background synchronisation task, to be able to propagate the organisation changes faster than periodic task. It means that we don't necessarily need "force reload" button in the interface.
For all the other cases (user does not exist, organisation token "not ok", etc) process remains same - plugin will show "Initialising plugin..." until the background task in successfully completed

## Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated

### CODE REVIEW

File: engine/apps/auth_token/auth.py [hunk 1]
```diff
@@ -96,6 +96,10 @@ def _get_user(request: Request, organization: Organization) -> User:
             logger.debug(f"Could not get user from grafana request. Context {context}")
             raise exceptions.AuthenticationFailed("Non-existent or anonymous user.")
 
+    @staticmethod
+    def get_user(cls, request: Request, organization: Organization) -> User:
```
Comment (Konstantinov-Innokentii) [inline engine/apps/auth_token/auth.py:100]: @staticmethod doesn't receive cls argument, probably you want to use @classmethod

File: engine/apps/grafana_plugin/views/sync.py [hunk 2]
```diff
@@ -25,12 +27,13 @@ def post(self, request: Request) -> Response:
 
         try:
             organization = Organization.objects.get(stack_id=stack_id, org_id=org_id)
-
-            if organization.api_token_status == Organization.API_TOKEN_STATUS_OK:
-                is_installed = True
-            organization.api_token_status = Organization.API_TOKEN_STATUS_PENDING
-
-            organization.save(update_fields=["api_token_status"])
+            try:
```
Comment (Konstantinov-Innokentii) [inline engine/apps/grafana_plugin/views/sync.py:30]: Could you please add comments here? I didn't catch the meaning of that check and why we setting API_TOKEN_STATUS_PENDING if we failed to get user from grafana context.

**Review by Konstantinov-Innokentii — CHANGES_REQUESTED** [review]

File: engine/apps/grafana_plugin/views/sync.py (hunk 2 above)
Comment (iskhakov) [reply engine/apps/grafana_plugin/views/sync.py]: For all the other cases (user does not exist, organisation token "not ok", etc) process remains same - plugin will show "Initialising plugin..." until the background task in successfully completed

File: engine/apps/auth_token/auth.py (hunk 1 above)
Comment (iskhakov) [reply engine/apps/auth_token/auth.py]: exactly, thanks!

File: engine/apps/grafana_plugin/views/sync.py (hunk 2 above)
Comment (Konstantinov-Innokentii) [reply engine/apps/grafana_plugin/views/sync.py]: Yep, but what is the case you are checking here and what is the expected behaviour? I mean, in which case we are unable to get user from grafana context?

Comment (iskhakov) [reply engine/apps/grafana_plugin/views/sync.py]: btw, cleaned up a little and added pr description
you were too fast:)

Comment (iskhakov) [reply engine/apps/grafana_plugin/views/sync.py]: fixed is_installed part, thanks

**Review by Konstantinov-Innokentii — APPROVED** [review]

File: engine/apps/grafana_plugin/views/sync.py [hunk 3]
```diff
@@ -22,26 +23,27 @@ def post(self, request: Request) -> Response:
...
         try:
             organization = Organization.objects.get(stack_id=stack_id, org_id=org_id)
-
             if organization.api_token_status == Organization.API_TOKEN_STATUS_OK:
                 is_installed = True
-            organization.api_token_status = Organization.API_TOKEN_STATUS_PENDING
 
-            organization.save(update_fields=["api_token_status"])
+            is_user = PluginAuthentication.is_user_from_request_present_in_organization(request, organization)
+            if not is_user:
```
Comment (joeyorlando) [inline engine/apps/grafana_plugin/views/sync.py:33]: _nit_;
```suggestion
            user_is_present_in_org = PluginAuthentication.is_user_from_request_present_in_organization(request, organization)
            if not user_is_present_in_org:
```

**Review by mderynck — APPROVED** [review]
LGTM
We can further optimize on UI side if we add the token_ok, license and version from the GET call to the data for the POST call. But that should probably be a different PR, this accomplishes most of the speedup while not needing to change UI.

**Comment by iskhakov** [comment]
Fixes part of this [issue](https://github.com/grafana/oncall/issues/1024)

**Pull request merged**